drivers/mipi_dbi/mipi_dbi_rpi_pico_pio.c

```diff
@@ -0,0 +1,535 @@
+	int pin_discarded;
+	const struct device *port;
+	struct mipi_dbi_pico_pio_sm sm;
+	struct mipi_dbi_pico_pio_dma dma;
+};
+
+struct mipi_dbi_pico_pio_data {
+	PIO pio;
+	struct k_mutex lock;
+	uint32_t sm_mask;
+	uint8_t split_count;
+	struct mipi_dbi_pico_pio_split split[MIPI_DBI_MAX_SPLITS];
+};
+
+static void mipi_dbi_pio_dma_handler(const struct device *dev, void *user_data, uint32_t channel,
+				     int status)
+{
+	const struct device *dev_mipi_dbi = (const struct device *)user_data;
+	const struct mipi_dbi_pico_pio_config *config = dev_mipi_dbi->config;
+	struct mipi_dbi_pico_pio_data *data = dev_mipi_dbi->data;
+
+	for (int i = 0; i < data->split_count; ++i) {
+		if (data->split[i].dma.channel == channel) {
+			k_msgq_put(config->msq_dma, &channel, K_NO_WAIT);
+		}
+	}
+}
+
+static bool mipi_dbi_pio_get_dma_slot(PIO pio, size_t sm, uint32_t *dma_slot)
+{
+	struct dma_slot_map {
+		PIO pio;
+		size_t sm;
+		uint32_t dma_slot;
+	};
+	static struct dma_slot_map dma_slots[NUM_PIO_STATE_MACHINES * NUM_PIOS] = {
+		/* clang-format off */
+		{pio0, 0, RPI_PICO_DMA_SLOT_PIO0_TX0},
+		{pio0, 1, RPI_PICO_DMA_SLOT_PIO0_TX1},
+		{pio0, 2, RPI_PICO_DMA_SLOT_PIO0_TX2},
+		{pio0, 3, RPI_PICO_DMA_SLOT_PIO0_TX3},
+		{pio1, 0, RPI_PICO_DMA_SLOT_PIO1_TX0},
+		{pio1, 1, RPI_PICO_DMA_SLOT_PIO1_TX1},
+		{pio1, 2, RPI_PICO_DMA_SLOT_PIO1_TX2},
+		{pio1, 3, RPI_PICO_DMA_SLOT_PIO1_TX3},
+#ifdef pio2
+		{pio2, 0, RPI_PICO_DMA_SLOT_PIO2_TX0},
+		{pio2, 1, RPI_PICO_DMA_SLOT_PIO2_TX1},
+		{pio2, 2, RPI_PICO_DMA_SLOT_PIO2_TX2},
+		{pio2, 3, RPI_PICO_DMA_SLOT_PIO2_TX3},
+#endif
+		/* clang-format on */
+	};
+	static_assert(ARRAY_SIZE(dma_slots) == NUM_PIOS * NUM_PIO_STATE_MACHINES);
+
+	for (int i = 0; i < NUM_PIO_STATE_MACHINES * NUM_PIOS; ++i) {
+		if (dma_slots[i].pio == pio && dma_slots[i].sm == sm) {
+			*dma_slot = dma_slots[i].dma_slot;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int mipi_dbi_pio_setup_dma(const struct device *dev, struct mipi_dbi_pico_pio_split *split)
+{
+	split->dma.channel = dma_claim_unused_channel(true);
+	if (split->dma.channel == -1) {
+		return -EPROTO;
+	}
+
+	struct mipi_dbi_pico_pio_data *data = dev->data;
+
+	uint32_t dma_slot;
+
+	if (!mipi_dbi_pio_get_dma_slot(data->pio, split->sm.sm, &dma_slot)) {
+		assert("no dma slot found for given pio and sm");
+		return -EPROTO;
+	}
+	split->dma.config.dma_slot = dma_slot;
+
+	/* memory to peripheral */
+	split->dma.config.channel_direction = 0b001;
+	split->dma.config.complete_callback_en = false;
+	split->dma.config.error_callback_dis = false;
+	/* either true == high, or false == default */
+	split->dma.config.channel_priority = false;
+	/* src and dest size have to be the same */
+	split->dma.config.source_data_size = 1;
+	split->dma.config.dest_data_size = split->dma.config.source_data_size;
+	split->dma.config.block_count = 1;
+	split->dma.head_block.source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+	/* not yet provided */
+	split->dma.head_block.source_address = 0;
```

Comment:
```suggestion
```

Same as with block size, it is filled right before dma_start, and no need in this line here. Correct me if I'm wrong